hotspots: Add welcome bot reply hotspot.

derAnfaenger opened the pull request:
Fixes #6030.

The hotspot doesn't look optimal to me yet. Ideally, it would pop up directly over the private message from the welcome bot, but I guess that's not how the hotspot system works.

My first migration :))

derAnfaenger commented:
I've updated a couple of the tests causing errors, yet there seems to be more to it. I'll try to make all tests pass once the basic appraoch here gets approved. Might need some help, however, especially for the `populate_db.py` thing (seems to fail nondeterministically for my changes)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -931,6 +931,26 @@ def do_send_messages(messages_maybe_none):
+                                              message['message'].sender,
+                                              ":tada: I'm out for now, but enjoy the rest of your Zulip experience!")
+                sender_hotspot_field.has_welcome_bot_replied_to_user = True
+                sender_hotspot_field.save()
```
@showell can help with this, but we want to avoid doing an extra database query in this codepath for every message we send.  There's a similar issue with the feedback bot that we're discussing how to resolve.  

Structurally, I think we want most of the logic here to go into a new function `handle_welcome_bot_reply` or something similar, maybe in `zerver/lib/onboarding.py` or a similar place.  Otherwise, we end up cluttering this code path a lot.

I would also move this if statement next to the feedback-bot one, since while the use of queuing is different, it's otherwise very analogous.

rishig commented:
Just took a look at this, sorry the issue description was unclear. We can store the state (i.e. whether or not welcome-bot has replied) in the hotspots table, but the welcome-bot reply should be a regular message, rather than a visual hotspot. Does that make sense?

derAnfaenger commented:
@rishig cool then, right now the PR covers both, but the problems seem to occur with the visual hotspot. I can remove it then :)

rishig commented:
One more thing: after talking it over with Tim, I think we actually don't want to use a hotspot for this. Could we just run a database query to check if it is your first PM to welcome-bot, and send the reply iff that's the case?

derAnfaenger commented:
Updated:
 * removed hotspot
* make usermessage query instead of storing a new has_replied flag.
* restructured, so that most of the logic resides in a separate method.

* found mypy bug (?):

I make the call `get_display_recipient(message['message'].recipient)[0]['email']`

 `get_display_recipient()` returns type `Union[Text, List[Dict[str, Any]]]`.

I got the error
> zerver/lib/actions.py:1069: error: Invalid index type "str" for "Union[str, Dict[str, Any]]"; expected type "Union[int, slice]"

so it seems to me that mypy is ignoring the Union and just interprets `Union[str, Dict[str, Any]]` as `str`?

derAnfaenger commented:
I'm still wondering though if the issue I found with mypy is actually a bug, so I can report it.

rishig commented:
Thanks @derAnfaenger for your work on this! Opened the other PR since ended up in a bit of a crunch.
A few things that were changed
* removed the edit to populate_db. One way to test this and general onboarding stuff is with `./manage.py add_new_realm`, and there is an open issue that should fix the general problem of populate_db realms being overly strange. Once you do `./manage.py add_new_realm`, you can use the new user that is created as a part of that process to manually test, and run `./manage.py add_new_user` to get additional users. 
* ended up moving the function from onboarding.py back into actions.py, to avoid the import loop.
* Used Message instead of UserMessage, since Message is a much smaller table in terms of number of rows, so queries are likely to be faster. Also, using something like message__sender causes a join query, which can be expensive.

If you have a chance between now and 12 hours from now, it would be great if tests could be added to #6734, to test 
* that the first pm to welcome bot gets exactly one response, and that the response is properly ordered (i.e. comes after the user's reply), and that the second pm to welcome bot doesn't get a response. 
* that sending a group-pm to welcome bot doesn't result in a response
* if you send a group-pm, and then send a pm, the pm does get a response

You may also need to fix some current failing tests on that PR, and address Tim's style comments. 

If not no worries, I can add them as well. If you do get a chance to do it, feel free to just push the new code to this PR and reopen it. Thanks!

derAnfaenger commented:
@rishig will try to add tests now :) 

I guess I didn't understand the purpose of `Message` correctly then? I thought it's a table containing all messages ever sent? Or is it only for the last n messages?

rishig commented:
awesome, thanks! 
hm .. good point. Old messages will indeed be in ArchivedMessage, which will break this. I guess the query should check both Message and ArchivedMessage? Good catch!

derAnfaenger commented:
@rishig updated this PR with your code :) 

I
* added a commit for including the welcome bot in tests
* added the tests for this feature (`test_tutorial.py` seemed the most reasonable place to do this
* fixed the style issue
* updated the query to check both `Message` and `ArchiveMessage`. Instead of counting the messages, I chose to check for the right message content, since that felt clearer and also makes the tests more straightforward (in the tests, a user does not receive an initial welcome message, thus altering the message count). I wonder what the performance issue is of querying `Message` and `ArchiveMessage` vs just querying `UserMessage`?

rishig commented:
great, thanks! The reason to use a count rather than the actual message, is that the actual message will change over time (e.g. at minimum it will be translated based on the default language of the realm).

rishig commented:
yeah, let's just do that.
So maybe, in populate_db:
* Create welcome-bot right after clear_database()
* Replace the Realm.objects.create(..) with something like
        realm = do_create_realm(..)
        setup_initial_streams(realm)
        iago = do_create_user(..)
        send_initial_pms(iago)
        setup_initial_private_stream(iago)
        send_initial_realm_messages(realm)
(eventually we should make a do_create_realm_with_initial_user function ..)
* Remove Iago from being created later
* Fix any tests that this breaks :).

timabbott commented:
Checking `ArchivedMessage` is a waste of time, in that those get deleted eventually.  It does suggest, though, that we should store this differently...

derAnfaenger commented:
Maybe we could fetch UserMessage and store a flag in the program so that the query only gets executed once per server startup?

rishig commented:
Yeah, it does suggest using a hotspot after all :/. I think this is fine
for now though, and we can make it more robust as follow-up?

On Sep 28, 2017 12:20 PM, "Robert Hönig" <notifications@github.com> wrote:

> Maybe we could fetch UserMessage and store a flag in the program so that
> the query only gets executed once per server startup?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/6424#issuecomment-332937679>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA2YH29VjE_X5odBvbVHD5m-lzrkPeb9ks5sm_F7gaJpZM4PQKgl>
> .
>

rishig commented:
Just talked with Tim, and the decision is we should just check the Message table for now, and eventually we can make this an embedded bot.

derAnfaenger commented:
@rishig I am not able to make `populate_db` follow the production flow:

The commit says that `do_create_realm` is dependent on the welcome bot. Yet I need an existing realm to create the welcome bot. Also, after checking `do_create_realm`, it rather seems to be dependent only on `NEW_USER_BOT`, with this snippet:
```python
# Send a notification to the admin realm (if configured)
    if settings.NEW_USER_BOT is not None:
        print("NEW_USER_BOT: ", settings.NEW_USER_BOT)
        signup_message = "Signups enabled"
        admin_realm = get_system_bot(settings.NEW_USER_BOT).realm
        internal_send_message(admin_realm, settings.NEW_USER_BOT, "stream",
                              "signups", string_id, signup_message)
```
`get_system_bot` errors in this case, even thought `settings.NEW_USER_BOT` is defined. So I tried the following:

```python
# Create our two default realms
            old_value = settings.NEW_USER_BOT
            settings.NEW_USER_BOT = None
            zulip_realm = do_create_realm(string_id="zulip", name="Zulip Dev", restricted_to_domain=True,
                invite_required=False, org_type=Realm.CORPORATE)
            settings.NEW_USER_BOT = old_value
            # Create welcome bot as the first required bot
            zulip_welcome_bot = [
                ("Welcome Bot", "welcome-bot@zulip.com"),
                ("New User Bot", "new-user-bot@zulip.com")
            ]
            create_users(zulip_realm, zulip_welcome_bot, bot_type=UserProfile.DEFAULT_BOT)
            #Execute skipped code in `do_create_realm`
            signup_message = "Signups enabled"
            admin_realm = get_system_bot(settings.NEW_USER_BOT).realm
            internal_send_message(admin_realm, settings.NEW_USER_BOT, "stream",
                              "signups", "zulip", signup_message)
            RealmDomain.objects.create(realm=zulip_realm, domain="zulip.com")
            setup_initial_streams(zulip_realm)
            # Create test Users (UserProfiles are automatically created,
            # as are subscriptions to the ability to receive personals).
            names = [
                ("Zoe", "ZOE@zulip.com"),
                ("Othello, the Moor of Venice", "othello@zulip.com"),
                ("Iago", "iago@zulip.com"),
                ("Prospero from The Tempest", "prospero@zulip.com"),
                ("Cordelia Lear", "cordelia@zulip.com"),
                ("King Hamlet", "hamlet@zulip.com"),
                ("aaron", "AARON@zulip.com"),
            ]
            for i in range(options["extra_users"]):
                names.append(('Extra User %d' % (i,), 'extrauser%d@zulip.com' % (i,)))
            create_users(zulip_realm, names)

            iago = get_user("iago@zulip.com", zulip_realm)
            do_change_is_admin(iago, True)
            iago.is_staff = True
            iago.save(update_fields=['is_staff'])
            send_initial_pms(iago)
            send_initial_realm_messages(zulip_realm)
```
This gives me a peculiar error I cannot trace back:
```
Traceback (most recent call last):
  File "/srv/zulip-py3-venv/lib/python3.4/random.py", line 253, in choice
    i = self._randbelow(len(seq))
  File "/srv/zulip-py3-venv/lib/python3.4/random.py", line 230, in _randbelow
    r = getrandbits(k)          # 0 <= r < 2**k
ValueError: number of bits must be greater than zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./manage.py", line 27, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/zulip-py3-venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-py3-venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 355, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-py3-venv/lib/python3.4/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip-py3-venv/lib/python3.4/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip/zilencer/management/commands/populate_db.py", line 271, in handle
    send_messages(job)
  File "/srv/zulip/zilencer/management/commands/populate_db.py", line 489, in send_messages
    recipient=message.recipient)).user_profile
  File "/srv/zulip-py3-venv/lib/python3.4/random.py", line 255, in choice
    raise IndexError('Cannot choose from an empty sequence')
IndexError: Cannot choose from an empty sequence
```

rishig commented:
interesting.

Possibly we should have a separate "system realm" that is created via do_create_realm, and where all these required bots live. Thoughts, @timabbott? It's also possible the right thing for now is just to merge this with the tests commented out, and then work on fixing that as followup.

showell commented:
I think the best long term plan is to just give each realm their own, dedicated system bots, so we can avoid lots of cross-realm yuckiness.  When we remove the restriction that emails can only be used once across all instances, that will possibly make that easier.

derAnfaenger commented:
So are there any updates I should make on this? For now, the test could also emulate the welcome message.

rishig commented:
Oh great, let's do that.

rishig commented:
How are things going on this? It's very close to done, so happy to just finish it off, unless you've been working on it offline.

derAnfaenger commented:
@rishig just updated to emulate the welcome message and change the database query so that only `Message` gets queried for the amount of messages.

rishig commented:
I'll take care of the conflicts.

rishig commented:
This looks great, thanks @derAnfaenger, and thanks for the quick turnaround for getting this into 1.7! Pushing this with minor spacing and comments changes to #6734. Also changed it to the more compact includes format:
`from foo import bar1, bar2`
rather than
```
from foo import (
    bar1,
    bar2,
)
```

rishig commented:
Closing, the code is now at #6734.
